feat: warn on element redefinitions, point to resolution docs

hunterloftis opened the pull request:
## Description

This is a working example of one path towards addressing versioning challenges for consumers.

Currently, SWC packages that are imported twice (either pre-built and included on the same page, or via different dependency versions in a single build) will fail at runtime with a custom element registration error.

This change converts that runtime error to a development-time warning, with the ability to direct consumers to documentation about resolving transitive dependency versions etc.

It also demonstrates how we can verify correct behavior in tests, via an example in sp-button (the tests are currently passing, but failing code coverage requirements since this adds a few lines of currently-uncovered code to each package).

The changes are a quick grep, so there may be extra packages to cover in a similar manner. Icons?

## Related issue(s)

- https://github.com/adobe/spectrum-web-components/issues/2666
- https://github.com/adobe/spectrum-web-components/discussions/2690
- https://github.com/adobe/spectrum-web-components/issues/864
- https://github.com/adobe/spectrum-web-components/discussions/2679
- https://github.com/adobe/spectrum-web-components/issues/1228

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [X] My code follows the code style of this project.
-   [X] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [X] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [X] I have added tests to cover my changes.
-   [X] All new and existing tests passed.
-   [X] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

hunterloftis commented:
@Westbrook what do you think re: this approach?

Westbrook reviewed (commented):
I really like the idea of leveraging the Dev Mode Messging System to support educating users about this situation. It makes we think we should add a `build` type and move this message to the `highest` level. Pairing it with content around possible fixes/workarounds in the link will be great.

I really DON'T like that this prevents the error. Without the error, it's hard to beleive any of our consumers does anything about this issue. Particularly, the team I've see (anecdotally) encounter this situation the most is specifically developing in a way that they should get this error until they actually get their architecture right. 

Structurally, for Dev Mode messages, the `if (window.__swc?.DEBUG) {` must be the outter most condition so that it can tree shake the largest amount of the work being done when built for production.

benjamind commented:
I agree with not suprressing the error, you only way you get folks to reliably fix things is to put roadblocks in their path to doing it the 'wrong' way.

Would welcome this dev message though, can't land soon enough in my opinion! :-)

hunterloftis commented:
@Westbrook @benjamind updated to still fail, but to inject a dev mode warning before failure

Also added a testing helper so we can be sure this is actually working in each component. I've converted all the "A" components so far.

Westbrook commented on packages/accordion/test/accordion-item.test.ts:
```diff
@@ -196,4 +197,8 @@ describe('Accordion Item', () => {
     });
+    describe(
+        'dev mode',
+        warnsOnDoubleRegister(() => import('../sp-accordion-item.js'))
```
Nice trick to load the second registration ourside of the export map!

Westbrook commented on packages/accordion/sp-accordion-item.ts:
```diff
@@ -10,8 +10,9 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 import { AccordionItem } from './src/AccordionItem.js';
+import { defineElement } from '@spectrum-web-components/shared/src/define-element.js';
```
Mochanical point that I could take or leave, not every package relies on `@spectrum-web-components/shared` but every package does rely on `@spectrum-web-components/base`, probably doesn't matter much (especially with the full path to code here), but wanted to make sure we agreed outloud this was a good place for this code to live.

hunterloftis commented on tools/base/src/define-element.ts:
```diff
@@ -0,0 +1,31 @@
+            window.__swc.warn(
+                undefined,
+                `Attempted to redefine <${name}>. This usually indicates that multiple versions of the same web component were loaded onto a single page.`,
+                'https://TODO'
```
First draft: https://dev-mode-define-element--spectrum-web-components.netlify.app/registry-conflicts/

Thoughts on what should go into yarn & pnpm?

hunterloftis commented on projects/documentation/content/registry-conflicts.md:
```diff
@@ -0,0 +1,54 @@
+---
```
Yep agreed. I started putting it with the other "guides" until I saw that the header for that section was "Contributing"

hunterloftis commented on projects/documentation/content/registry-conflicts.md:
```diff
@@ -0,0 +1,54 @@
+
+### With yarn classic
+
+### With pnpm
```
Hey @benjamind can you confirm or deny that `pnpm dedupe` hoists as expected? Maybe by intentionally shifting a transitive dep in one of your projects to something lower-but-compatible, then deduping that conflict away?

hunterloftis commented on projects/documentation/content/registry-conflicts.md:
```diff
@@ -0,0 +1,54 @@
+
+### With pnpm
+
+## Future
```
I'm down with a CTA if you have one in mind! I'll note the polyfill performance results.

Westbrook replied:
Maybe something like:
> To support this feature in coming to browsers faster, [share you use case](https://github.com/WICG/webcomponents/issues/716) with implementors to help increase awareness and priority of this API.

Westbrook commented on projects/documentation/content/registry-conflicts.md:
```diff
@@ -0,0 +1,64 @@
+
+-   Running `yarn list {packagename}` to see which versions are installed
+-   Adding a commonly-valid version to [yarn's resolutions list](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) in package.json.
+-   Running `rm yarn.lock && yarn` to reinstall with resolutions.
```
If you add `resolutions` you _shouldn't_ need to nuke the lock, right?

Westbrook commented on projects/documentation/content/registry-conflicts.md:
```diff
@@ -0,0 +1,64 @@
+
+## Resolution
+
+First, verify that each `@spectrum-web-components` package in your `package.json` shares the same version. If `@spectrum-web-components/button` is on version `0.4.0`, then all other SWC packages should _also_ be on version `0.4.0`.
```
We know that the first version that supports this is `0.30.0` should we use that rather than one we know doesn't support this?

hunterloftis replied:
Good call

hunterloftis commented:
Closing to retain work history, shifting PR to https://github.com/adobe/spectrum-web-components/pull/3171
